### PR TITLE
Input arguments as kwargs

### PR DESCRIPTION
**Pull request by nfarabullini**
Allow for input keyword arguments to be used in function calls, e.g.
```python
@program
def program_input_kwargs(
    a: Field[[IDim, JDim], float64],
    b: Field[[IDim, JDim], float64],
    c: Field[[IDim, JDim], float64],
    out: Field[[IDim, JDim], float64],
):
    fieldop_input_kwargs(c, a, b, out=out)
program_input_kwargs(a=input_1, b=input_2, c=input_3, out=out, offset_provider={})
```

### CODE REVIEW

File: src/functional/ffront/decorator.py [hunk 1]
```diff
@@ -260,6 +260,36 @@ def format_itir(
...
+        new_args_ls = [None] * len(past_params)
+        wrong_args = [False] * len(past_params)
+        kwargs_count = 0
+        if len(past_params) < len(args) + len(kwargs):
+            raise GTTypeError(
+                "Number of parameters in function call exceeds number in function definition."
+            )
+        for param_i, param in enumerate(past_params):
+            if param.id in kwargs:
+                new_args_ls[param_i] = kwargs[param.id]
```
Comment (tehrengruber) [inline src/functional/ffront/decorator.py:275]:
```suggestion
                if param_i+1 >= len(args):
                  raise ProgramTypeError.from_past_node(
                      self.past_node, f"got multiple values for argument 'param.id''"
                  )
                new_args_ls[param_i] = kwargs[param.id]
```

For reference:
```
>>> def func(a):
...   return a
... 
>>> func(1, a=1)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: func() got multiple values for argument 'a'
>>>
```

File: src/functional/ffront/decorator.py [hunk 2]
```diff
@@ -260,6 +260,36 @@ def format_itir(
...
+                kwargs.pop(param.id)
+                kwargs_count += 1
+            elif len(args) > param_i - kwargs_count:
+                new_args_ls[param_i] = args[param_i - kwargs_count]
```
Comment (tehrengruber) [inline src/functional/ffront/decorator.py:279]:
```suggestion
            elif param_i+1 >= len(args):
                new_args_ls[param_i] = args[param_i]
```
The idea behind having `new_args_ls` was to avoid avoid exactly this kind of index computations for the access, which is not correct here. Surprisingly the tests don't fail, not exactly sure why. If you consider such a case:
```python
def func(a, b, c, d):
    ...
func(a, b, c, d=1)
```
You execute the following:
```python
new_args_ls[0] = args[0 - 1]
new_args_ls[1] = args[1 - 1]
```

**Review by tehrengruber — CHANGES_REQUESTED** [review]
Just a partial review. It became too complicated to add more suggestions.

File: src/functional/ffront/decorator.py (hunk 2 above)
Comment (nfarabullini) [reply src/functional/ffront/decorator.py]: this is not correct, we can discuss on Monday

File: src/functional/ffront/decorator.py (hunk 1 above)
Comment (nfarabullini) [reply src/functional/ffront/decorator.py:279]: this cannot work, we can discuss on Monday

File: tests/functional_tests/ffront_tests/test_execution.py [hunk 3]
```diff
@@ -1271,3 +1271,110 @@ def test_undefined_symbols():
         @field_operator
         def return_undefined():
             return undefined_symbol
+
+
+def test_input_kwargs_1(fieldview_backend):
```
[the remark this answers is not in the document]
Comment (nfarabullini) [reply tests/functional_tests/ffront_tests/test_execution.py:1276]: I am ok moving them but I disagree to merge them. It is not clean

File: src/functional/ffront/decorator.py (hunk 1 above)
Comment (nfarabullini) [reply src/functional/ffront/decorator.py]: how about this: https://github.com/GridTools/gt4py/pull/982/commits/0873010659d5d5ec3a8db3f1b76f436bd5fc8db3

File: src/functional/ffront/decorator.py [hunk 4]
```diff
@@ -94,6 +94,26 @@ def _filter_closure_vars_by_type(closure_vars: dict[str, Any], *types: type) ->
...
+        if param.id in kwargs:
+            if param.id not in ["out", "domain"] and param_i < len(args):
+                raise ProgramTypeError(f"got multiple values for argument {param.id}.")
+            new_args.append(kwargs[param.id])
+            kwargs.pop(param.id)
+        elif param_i < len(args):
+            new_args.append(args[param_i])
+        else:
+            # case when parameter is in function definition but not in function call
+            # e.g. function definition includes 3 parameters, but only 2 were given as inputs
```
Comment (tehrengruber) [inline src/functional/ffront/decorator.py:110]:
```suggestion
            # e.g. function expects 3 parameters, but only 2 were given. To avoid duplication of
            # error handling this is covered later in `accept_args`.
```
Nitpicking.

File: src/functional/ffront/decorator.py [hunk 5]
```diff
@@ -94,6 +94,26 @@ def _filter_closure_vars_by_type(closure_vars: dict[str, Any], *types: type) ->
     return {name: value for name, value in closure_vars.items() if isinstance(value, types)}
 
 
+def _canonicalize_args(node_params, args, kwargs) -> tuple[tuple, dict]:
+    new_args = []
+
+    for param_i, param in enumerate(node_params):
+        if param.id in kwargs:
+            if param.id not in ["out", "domain"] and param_i < len(args):
```
Comment (tehrengruber) [inline src/functional/ffront/decorator.py:102]: For my own understanding why are `out` and `domain` excluded?

File: src/functional/ffront/decorator.py [hunk 6]
```diff
@@ -94,6 +94,26 @@ def _filter_closure_vars_by_type(closure_vars: dict[str, Any], *types: type) ->
...
+                raise ProgramTypeError(f"got multiple values for argument {param.id}.")
+            new_args.append(kwargs[param.id])
+            kwargs.pop(param.id)
```
Comment (tehrengruber) [inline src/functional/ffront/decorator.py:105]:
```suggestion
```
To avoid modifying state in the caller. Isn't necessary anyway right?

File: src/functional/ffront/decorator.py [hunk 7]
```diff
@@ -281,11 +301,8 @@ def format_itir(
         )
 
     def _validate_args(self, *args, **kwargs) -> None:
-        if kwargs:
-            raise NotImplementedError("Keyword arguments are not supported yet.")
```
Comment (tehrengruber) [inline src/functional/ffront/decorator.py:285]:
```
    raise NotImplementedError("Keyword-only arguments are not supported yet.")
```

File: src/functional/ffront/decorator.py [hunk 8]
```diff
@@ -281,11 +301,8 @@ def format_itir(
...
-
         arg_types = [symbol_makers.make_symbol_type_from_value(arg) for arg in args]
-        kwarg_types = {k: symbol_makers.make_symbol_type_from_value(v) for k, v in kwargs.items()}
+        kwarg_types = {}
```
Comment (tehrengruber) [inline src/functional/ffront/decorator.py:305]:
```suggestion
        kwarg_types = {k: symbol_makers.make_symbol_type_from_value(v) for k, v in kwargs.items()}
```

File: src/functional/ffront/decorator.py [hunk 9]
```diff
@@ -300,6 +317,10 @@ def _validate_args(self, *args, **kwargs) -> None:
             ) from err
 
     def _process_args(self, args: tuple, kwargs: dict) -> tuple[tuple, tuple, dict[str, Any]]:
+        # if parameter is in signature but not in args, move it from kwargs to args
+        if len(kwargs) > 0:
+            args, kwargs = _canonicalize_args(self.past_node.params, args, kwargs)
```
Comment (tehrengruber) [inline src/functional/ffront/decorator.py:322]:
```suggestion
        args, kwargs = _canonicalize_args(self.past_node.params, args, kwargs)
```
Either skip the check or move it to `_canonicalize_args`.

File: src/functional/ffront/decorator.py [hunk 10]
```diff
@@ -556,6 +577,8 @@ def __call__(
         offset_provider: dict[str, Dimension],
         **kwargs,
     ) -> None:
+
```
Comment (tehrengruber) [inline src/functional/ffront/decorator.py:580]:
```suggestion
```

File: tests/functional_tests/ffront_tests/test_program.py [hunk 11]
```diff
@@ -278,3 +278,43 @@ def empty_domain_program(
...
+    @program
+    def program_input_kwargs(
+        a: Field[[IDim, JDim], float64],
+        b: Field[[IDim, JDim], float64],
+        c: Field[[IDim, JDim], float64],
+        out: Field[[IDim, JDim], float64],
+    ):
+        fieldop_input_kwargs(c, a, b, out=out)
+
+    fieldop_input_kwargs(input_1, b=input_2, c=input_3, out=out, offset_provider={})
```
Comment (tehrengruber) [inline tests/functional_tests/ffront_tests/test_program.py:309]:
```suggestion
    fieldop_input_kwargs(input_1, b=input_2, c=input_3, out=out, offset_provider={})
    assert np.allclose(expected, out)
    out[:] = 0
```
Same below.

File: src/functional/ffront/decorator.py [hunk 12]
```diff
@@ -94,6 +94,26 @@ def _filter_closure_vars_by_type(closure_vars: dict[str, Any], *types: type) ->
     return {name: value for name, value in closure_vars.items() if isinstance(value, types)}
 
 
+def _canonicalize_args(node_params, args, kwargs) -> tuple[tuple, dict]:
```
Comment (tehrengruber) [inline src/functional/ffront/decorator.py:97]: Type annotations for the arguments.

File: src/functional/ffront/decorator.py [hunk 13]
```diff
@@ -94,6 +94,26 @@ def _filter_closure_vars_by_type(closure_vars: dict[str, Any], *types: type) ->
...
+    new_args = []
```
Comment (tehrengruber) [inline src/functional/ffront/decorator.py:98]: Type annotation.

**Review by tehrengruber — CHANGES_REQUESTED** [review]
As discussed.

File: src/functional/ffront/decorator.py (hunk 6 above)
Comment (nfarabullini) [reply src/functional/ffront/decorator.py:105]: if you don't pop the arguments, you the error in the comment below is triggered

File: src/functional/ffront/decorator.py (hunk 5 above)
Comment (nfarabullini) [reply src/functional/ffront/decorator.py:102]: e.g. when you call: ```program(input_1, b=input_2, c=input_3, out=out, offset_provider={})````

In this case out is also a keyword argument

File: src/functional/ffront/decorator.py (hunk 6 above)
Comment (tehrengruber) [reply src/functional/ffront/decorator.py:105]: I see, instead of workings with `kwargs` use a copy, e.g.
```
new_kwargs = {**kwargs}
```

File: src/functional/ffront/decorator.py [hunk 14]
```diff
@@ -300,6 +323,9 @@ def _validate_args(self, *args, **kwargs) -> None:
             ) from err
 
     def _process_args(self, args: tuple, kwargs: dict) -> tuple[tuple, tuple, dict[str, Any]]:
+        # if parameter is in signature but not in args, move it from kwargs to args
```
Comment (tehrengruber) [inline src/functional/ffront/decorator.py:326]:
```suggestion
```

File: src/functional/ffront/decorator.py [hunk 15]
```diff
@@ -94,6 +94,29 @@ def _filter_closure_vars_by_type(closure_vars: dict[str, Any], *types: type) ->
...
+        if param.id in kwargs:
+            if param.id not in ["out", "domain"] and param_i < len(args):
+                raise ProgramTypeError(f"got multiple values for argument {param.id}.")
+            new_args.append(kwargs[param.id])
+            kwargs.pop(param.id)
+        elif param_i < len(args):
+            new_args.append(args[param_i])
+        else:
+            # case when parameter is in function definition but not in function call
+            # e.g. function expects 3 parameters, but only 2 were given. To avoid duplication of
```
Comment (tehrengruber) [inline src/functional/ffront/decorator.py:112]: Format.

File: tests/functional_tests/ffront_tests/test_program.py [hunk 16]
```diff
@@ -278,3 +278,48 @@ def empty_domain_program(
...
+        c: Field[[IDim, JDim], float64],
+        out: Field[[IDim, JDim], float64],
+    ):
+        fieldop_input_kwargs(c, a, b, out=out)
+
+    out = np_as_located_field(IDim, JDim)(np.zeros((size, size)))
+
+    program_input_kwargs(input_1, b=input_2, c=input_3, out=out, offset_provider={})
+    assert np.allclose(expected_program, out)
+    out = np_as_located_field(IDim, JDim)(np.zeros((size, size)))
```
Comment (tehrengruber) [inline tests/functional_tests/ffront_tests/test_program.py:317]:
```suggestion
    out = np_as_located_field(IDim, JDim)(np.zeros((size, size)))
    program_input_kwargs(input_1, b=input_2, c=input_3, out=out, offset_provider={})
    assert np.allclose(expected_program, out)
```
If you reinitialize the out field please use this grouping everywhere.

File: tests/functional_tests/ffront_tests/test_program.py [hunk 17]
```diff
@@ -278,3 +278,48 @@ def empty_domain_program(
...
+    assert np.allclose(expected_fieldop, out)
+
+    @program
+    def program_input_kwargs(
+        a: Field[[IDim, JDim], float64],
+        b: Field[[IDim, JDim], float64],
+        c: Field[[IDim, JDim], float64],
+        out: Field[[IDim, JDim], float64],
+    ):
+        fieldop_input_kwargs(c, a, b, out=out)
```
Comment (tehrengruber) [inline tests/functional_tests/ffront_tests/test_program.py:311]:
```suggestion
        fieldop_input_kwargs(a, b, c, out=out)
```
What is the point of this argument swap? Testing if calls inside programs work should not be the objective of this tests.

File: tests/functional_tests/ffront_tests/test_program.py [hunk 18]
```diff
@@ -278,3 +278,48 @@ def empty_domain_program(
...
+    out = np_as_located_field(IDim, JDim)(np.zeros((size, size)))
+
+    program_input_kwargs(input_1, b=input_2, c=input_3, out=out, offset_provider={})
+    assert np.allclose(expected_program, out)
+    out = np_as_located_field(IDim, JDim)(np.zeros((size, size)))
+
+    program_input_kwargs(a=input_1, b=input_2, c=input_3, out=out, offset_provider={})
+    assert np.allclose(expected_program, out)
+
+    with pytest.raises(GTTypeError) as exc_info_1:
```
Comment (tehrengruber) [inline tests/functional_tests/ffront_tests/test_program.py:322]: Since this is not a multi-level-exception (or however they are I called, I always forget, I mean the `raise ... from e` where one is interested in the message in `e`), you can use the regular pytest pattern. Here is the excerpt from the [docs](https://docs.pytest.org/en/7.1.x/how-to/assert.html):
```python
def test_match():
    with pytest.raises(ValueError, match=r".* 123 .*"):
        myfunc()
```

**Review by tehrengruber — CHANGES_REQUESTED** [review]

File: src/functional/ffront/decorator.py (hunk 5 above)
Comment (nfarabullini) [reply src/functional/ffront/decorator.py:104]: I did some testing, you are right

File: src/functional/ffront/decorator.py (hunk 15 above)
Comment (tehrengruber) [reply src/functional/ffront/decorator.py:113]: This is not resolved, please fix the formatting.

File: src/functional/ffront/decorator.py [hunk 19]
```diff
@@ -94,6 +94,30 @@ def _filter_closure_vars_by_type(closure_vars: dict[str, Any], *types: type) ->
     return {name: value for name, value in closure_vars.items() if isinstance(value, types)}
 
 
+def _canonicalize_args(
+    node_params: list[ct.SymbolType], args: tuple[ct.SymbolType], kwargs: dict[str, ct.SymbolType]
```
Comment (tehrengruber) [inline src/functional/ffront/decorator.py:98]: The `list[ct.SymbolType]` annotation is not correct I believe.

**Review by tehrengruber — CHANGES_REQUESTED** [review]

File: src/functional/ffront/decorator.py (hunk 19 above)
Comment (nfarabullini) [reply src/functional/ffront/decorator.py:98]: it is

Comment (nfarabullini) [reply src/functional/ffront/decorator.py:98]: it is

File: tests/functional_tests/ffront_tests/test_program.py [hunk 20]
```diff
@@ -278,3 +279,45 @@ def empty_domain_program(
...
     ):
         empty_domain_program(a, out_field, offset_provider={})
+
+
+def test_input_kwargs(fieldview_backend):
+    size = 10
+    input_1 = np_as_located_field(IDim, JDim)(np.ones((size, size)))
+    input_2 = np_as_located_field(IDim, JDim)(np.ones((size, size)) * 2)
+    input_3 = np_as_located_field(IDim, JDim)(np.ones((size, size)) * 3)
+    out = np_as_located_field(IDim, JDim)(np.zeros((size, size)))
```
Comment (tehrengruber) [inline tests/functional_tests/ffront_tests/test_program.py:289]: Superfluous by now.

File: src/functional/ffront/decorator.py [hunk 21]
```diff
@@ -94,6 +94,32 @@ def _filter_closure_vars_by_type(closure_vars: dict[str, Any], *types: type) ->
...
+            if param_i < len(args):
+                raise ProgramTypeError(f"got multiple values for argument {param.id}.")
+            new_args.append(kwargs[param.id])
+            new_kwargs.pop(param.id)
+        elif param_i < len(args):
+            new_args.append(args[param_i])
+        else:
+            # case when param in function definition but not in function call
+            # e.g. function expects 3 parameters, but only 2 were given.
+            # Error covered later in `accept_args`.
```
Comment (tehrengruber) [inline src/functional/ffront/decorator.py:116]: Why did you remove the part of the sentence that motivates why errors are not covered here?

**Review by tehrengruber — CHANGES_REQUESTED** [review]

File: src/functional/ffront/decorator.py (hunk 21 above)
Comment (nfarabullini) [reply src/functional/ffront/decorator.py:116]: I merely re-phrased, it still says: "Error covered later in `accept_args`."

**Review by tehrengruber — APPROVED** [review]
